url7a69e77641_benmouh_MySgbd_tgz-pJ8-Demo_CheckRowNumJ8/src/org/sgbd/Mysgbd/data/MEval.java

```diff
@@ -0,0 +1,317 @@
+
+      if(op.equals("NOT IN"))
+        return ! eval(tuple, newexp);
+      else
+        return eval(tuple, newexp);
+
+    } else if(op.equals("IS NULL")) {
+
+      if(pred.nbOperands() <= 0 || pred.getOperand(0) == null) return true;
+      MExp x = pred.getOperand(0);
+      if(x instanceof MConstant) {
+        return (((MConstant)x).getType() == MConstant.NULL);
+      } else {
+        throw new SQLException("MEval.eval(): can't eval IS (NOT) NULL");
+      }
+
+    } else if(op.equals("IS NOT NULL")) {
+
+      MExpression x = new MExpression("IS NULL");
+      x.setOperands(pred.getOperands());
+      return ! eval(tuple, x);
+
+    } else {
+      throw new SQLException("MEval.eval(): Unknown operator " + op);
+    }
+
+  }
+
+  double evalCmp(MTuple tuple, Vector operands) throws SQLException {
+
+    if(operands.size() < 2) {
+      throw new SQLException(
+        "MEval.evalCmp(): Trying to compare less than two values");
+    }
+    if(operands.size() > 2) {
+      throw new SQLException(
+        "MEval.evalCmp(): Trying to compare more than two values");
+    }
+
+    Object o1 = null, o2 = null, obj;
+
+    o1 = evalExpValue(tuple, (MExp)operands.elementAt(0));
+    o2 = evalExpValue(tuple, (MExp)operands.elementAt(1));
+
+    if(o1 instanceof String || o2 instanceof String) {
+      return(o1.equals(o2) ? 0 : -1);
+    }
+
+    if(o1 instanceof Number && o2 instanceof Number) {
+      return ((Number)o1).doubleValue() - ((Number)o2).doubleValue();
+    } else {
+      throw new SQLException("MEval.evalCmp(): can't compare (" + o1.toString()
+        + ") with (" + o2.toString() + ")");
+    }
+  }
+
+
+  // -------------------------------------------------------------------------
+  /**
+   * evalLike
+   * evaluates the LIKE operand
+   * 
+   * @param tuple the tuple to evaluate
+   * @param operands the operands
+   * @return true-> the expression matches
+   * @throws SQLException 
+   */
+  private boolean evalLike(MTuple tuple, Vector operands) throws SQLException
+  {
+    if(operands.size() < 2) {
+      throw new SQLException(
+        "MEval.evalCmp(): Trying to compare less than two values");
+    }
+    if(operands.size() > 2) {
+      throw new SQLException(
+        "MEval.evalCmp(): Trying to compare more than two values");
+    }
+
+    Object o1 = evalExpValue(tuple, (MExp)operands.elementAt(0));
+    Object o2 = evalExpValue(tuple, (MExp)operands.elementAt(1));
+
+    if ( (o1 instanceof String) && (o2 instanceof String) ) {
+      String s1 = (String)o1;
+      String s2 = (String)o2;
+      if ( s2.startsWith("%") ) {
+        return s1.endsWith(s2.substring(1));
+      } else if ( s2.endsWith("%") ) {
+        return s1.startsWith(s2.substring(0,s2.length()-1));
+      } else {
+        return s1.equalsIgnoreCase(s2);
+      }
+    }
+    else {
+      throw new SQLException("MEval.evalLike(): LIKE can only compare strings");
+    }
+
+  }
+
+  double evalNumericExp(MTuple tuple, MExpression exp)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 16. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.